pkg/controllers/consolidation/controller.go

```diff
@@ -0,0 +1,665 @@
+		if err := c.launchReplacementNode(ctx, action); err != nil {
+			// If we failed to launch the replacement, don't consolidate.  If this is some permanent failure,
+			// we don't want to disrupt workloads with no way to provision new nodes for them.
+			logging.FromContext(ctx).Errorf("Launching replacement node, %s", err)
+			return
+		}
+	}
+
+	for _, oldNode := range action.oldNodes {
+		c.recorder.TerminatingNodeForConsolidation(oldNode, action.String())
+		if err := c.kubeClient.Delete(ctx, oldNode); err != nil {
+			logging.FromContext(ctx).Errorf("Deleting node, %s", err)
+		} else {
+			consolidationNodesTerminatedCounter.Add(1)
+		}
+	}
+}
+
+func byNodeDisruptionCost(nodes []candidateNode) func(i int, j int) bool {
+	return func(a, b int) bool {
+		if nodes[a].disruptionCost == nodes[b].disruptionCost {
+			// if costs are equal, choose the older node
+			return nodes[a].CreationTimestamp.Before(&nodes[b].CreationTimestamp)
+		}
+		return nodes[a].disruptionCost < nodes[b].disruptionCost
+	}
+}
+
+// launchReplacementNode launches a replacement node and blocks until it is ready
+func (c *Controller) launchReplacementNode(ctx context.Context, minCost consolidationAction) error {
+	defer metrics.Measure(consolidationReplacementNodeInitializedHistogram)()
+	if len(minCost.oldNodes) != 1 {
+		return fmt.Errorf("expected a single node to replace, found %d", len(minCost.oldNodes))
+	}
+
+	// cordon the node before we launch the replacement to prevent new pods from scheduling to the node
+	if err := c.setNodeUnschedulable(ctx, minCost.oldNodes[0].Name, true); err != nil {
+		return fmt.Errorf("cordoning node %s, %w", minCost.oldNodes[0].Name, err)
+	}
+
+	nodeNames, err := c.provisioner.LaunchNodes(ctx, provisioning.LaunchOptions{RecordPodNomination: false}, minCost.replacementNode)
+	if err != nil {
+		return err
+	}
+	if len(nodeNames) != 1 {
+		return fmt.Errorf("expected a single node name, got %d", len(nodeNames))
+	}
+
+	consolidationNodesCreatedCounter.Add(1)
+
+	var k8Node v1.Node
+	// Wait for the node to be ready, we can wait for up to about 3 minutes which is ok as all this blocks is
+	// further consolidation.
+	var once sync.Once
+	if err := retry.Do(func() error {
+		if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodeNames[0]}, &k8Node); err != nil {
+			return fmt.Errorf("getting node, %w", err)
+		}
+		once.Do(func() {
+			c.recorder.LaunchingNodeForConsolidation(&k8Node, minCost.String())
+		})
+
+		if _, ok := k8Node.Labels[v1alpha5.LabelNodeInitialized]; !ok {
+			// make the user aware of why consolidation is paused
+			c.recorder.WaitingOnReadinessForConsolidation(&k8Node)
+			return errors.New("node is not initialized")
+		}
+		return nil
+	}, retry.Delay(2*time.Second),
+		retry.LastErrorOnly(true),
+		retry.Attempts(30),
+		retry.MaxDelay(10*time.Second), // ~ 4.5 minutes in total
+	); err != nil {
+		// node never become ready, so uncordon the node we were trying to delete and report the error
```

Comment:
Can we check that the nodeDeletionTimestamp is not nil here before we uncordon? Worried about the case where a user decides to `kubectl delete node` a node and we uncordon it, allowing pods to schedule while we're trying to drain it.